File: client/common/serialization.go
package common

import "bytes"

// SBDSerialization (Sequential Binary Data Serialization) provides
// binary serialization and deserialization for a list of ordered strings.

// Serialize a sequence of strings into a sequential binary strings
// bytes representation.
func SBDSerialize(parts []string) []byte {
	data := make([]byte, 0)

	for _, part := range parts {
		data = append(data, []byte(part)...)
		data = append(data, []byte{0}...)
	}

	// Remove the last 0 byte
	if len(data) > 0 {
		data = data[:len(data)-1]
	}

	return data
}

func SBDSerializeArray(items [][]string) []byte {
	data := make([]byte, 0)

	for _, item := range items {
		data = append(data, SBDSerialize(item)...)
		data = append(data, []byte{0}...)
	}

	// Remove the last 0 byte
	if len(data) > 0 {
		data = data[:len(data)-1]
	}

	return data
}

func SBDDeserialize(data []byte) []string {
	parts := bytes.Split(data, []byte{0})
	strings := make([]string, len(parts))

	for i, part := range parts {
		strings[i] = string(part)
	}

	return strings
}

File: server/common/serialization.py
from typing import Optional

class SBDSerialization:
  """
  SBDSerialization (Sequential Binary Data Serialization) provides
  binary serialization and deserialization for a list of ordered strings.
  """

  def serialize(parts: 'list[str]') -> bytes:
    """
    Serialize an ordered list of strings into a sequence of binary data
    bytes representation.
    """

    return b'\0'.join(part.encode('utf-8') for part in parts)
  
  def serialize_array(parts: 'list[str]') -> bytes:
    """
    Serialize a list of ordered list strings into a sequence of binary data
    bytes representation.
    """

    return b'\0'.join(SBDSerialization.serialize(part) for part in parts)

  def deserialize(data: 'bytes', expected_parts: 'Optional[int]') -> 'list[str]':
    """
    Deserialize a sequence of binary data bytes representation
    into an ordered list of strings.

    If expected_parts is not None, it will check if the number of
    elements in the data matches the expected number of elements.
    """

    parts = data.split(b'\0')
  
    if expected_parts is not None and len(parts) != expected_parts:
      raise ValueError(f'Invalid data format ({data}). Expected {expected_parts} elements, got {len(parts)}.')

    return [part.decode('utf-8') for part in parts]

  def deserialize_to_object(data: 'bytes', structure: 'list[str]') -> 'object':
    """
    Deserialize a sequential binary strings bytes representation
    into an object with the given ordered structure.
    """

    parts = SBDSerialization.deserialize(data, len(structure))

    return { structure[i]: parts[i] for i in range(len(parts)) }

  def deserialize_array(data: 'bytes', expected_parts: int) -> 'list[object]':
    """
    Deserialize an array of sequential binary strings bytes representation
    into a list.
    """

    parts = SBDSerialization.deserialize(data, None)

    if expected_parts is not None and len(parts) % expected_parts != 0:
      raise ValueError(f'Invalid data format ({data}). Expected items of {expected_parts} parts, got {len(parts) / expected_parts} (approximately).')

    return [parts[i:i + expected_parts] for i in range(0, len(parts), expected_parts)]